Book.java

```diff
@@ -0,0 +1,315 @@
+package accounting;
+
+import agents.Agent;
+import actions.Action;
+import contracts.Asset;
+import contracts.Contract;
+import contracts.Loan;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+
+/**
+ * This is the main class implementing double entry accounting. All public operations provided by this class
+ * are performed as a double entry operation, i.e. a pair of (dr, cr) operations.
+ *
+ * A Book contains a set of accounts, and is the interface between an agent and its accounts. Agents cannot
+ * directly interact with accounts other than via a Book.
+ *
+ * At the moment, a Book contains an account for each type of contract, plus an equity account and a cash account.
+ *
+ * A simple economic agent will usually have a single Book, whereas complex firms and banks can have several books
+ * (as in branch banking for example).
+ *
+ * @author rafa
+ */
+public class Book implements BookAPI {
+
+    public Book(Agent owner) {
+        // A Book is a list of accounts
+        accounts = new HashSet<>();
+
+        // Subsets of the list of accounts (for quicker searching)
+        assetAccounts = new HashSet<>();
+        liabilityAccounts = new HashSet<>();
+        equityAccounts = new HashSet<>();
+
+        // Each Account includes an inventory to hold one type of contract.
+        // These hashmaps are used to access the correct account for a given type of contract.
+        // Note that separate hashmaps are needed for asset accounts and liability accounts: the same contract
+        // type (such as Loan) can sometimes be an asset and sometimes a liability.
+        assetAccountsMap = new HashMap<>();
+        liabilityAccountsMap = new HashMap<>();
+
+        // Not sure whether I should be passing the owner...
```

Comment:
I think this is a good idea. There may be times when the book realizes that something requires the owner's attention. With this architecture, the book can ping the owner if required.